sdk/lib/rtl/path.c

```diff
@@ -459,12 +464,130 @@ RtlGetLengthWithoutLastFullDosOrNtPathElement(IN ULONG Flags,
+    {
+        /* Just keep the dll name, ignore path components */
+        Position += sizeof(WCHAR);
+        DllNameOnly.Buffer = DllName->Buffer + Position / sizeof(WCHAR);
+        DllNameOnly.Length = DllName->Length - Position;
+        DllNameOnly.MaximumLength = DllName->MaximumLength - Position;
+    }
+
+    if (!NT_SUCCESS(RtlFindCharInUnicodeString(RTL_FIND_CHAR_IN_UNICODE_STRING_START_AT_END,
+                                               &DllNameOnly, &ExtensionChar, &Position)))
+    {
+        Position = 0;
+    }
+
+    HasExtension = Position > 1;
+
+    /* First we create the c:\path\processname.exe.Local\something.dll path */
+    RequiredSize = ImagePathName.Length + RtlpDotLocal.Length + DllNameOnly.Length +
+        (HasExtension ? 0 : RtlpDefaultExtension.Length) + sizeof(UNICODE_NULL);
+
+    /* This is not going to work out */
+    if (RequiredSize >= UNICODE_STRING_MAX_BYTES)
+        return STATUS_NAME_TOO_LONG;
+
+    /* We need something extra */
+    if (RequiredSize > CopyLocalName.MaximumLength)
+    {
+        CopyLocalName.Buffer = RtlpAllocateStringMemory(RequiredSize, TAG_USTR);
+        if (CopyLocalName.Buffer == NULL)
+            return STATUS_NO_MEMORY;
+        CopyLocalName.MaximumLength = RequiredSize;
+    }
+    /* Now build the entire path */
+    CopyLocalName.Length = 0;
+    Status = RtlAppendUnicodeStringToString(&CopyLocalName, &ImagePathName);
+    if NT_SUCCESS(Status)
+        Status = RtlAppendUnicodeStringToString(&CopyLocalName, &RtlpDotLocal);
+    if NT_SUCCESS(Status)
+        Status = RtlAppendUnicodeStringToString(&CopyLocalName, &DllNameOnly);
+    /* Do we need to append an extension? */
+    if (NT_SUCCESS(Status) && !HasExtension)
+        Status = RtlAppendUnicodeStringToString(&CopyLocalName, &RtlpDefaultExtension);
+
+    if (NT_SUCCESS(Status))
+    {
+        /* then we create the c:\path\something.dll path */
+        if (NT_SUCCESS(RtlFindCharInUnicodeString(RTL_FIND_CHAR_IN_UNICODE_STRING_START_AT_END,
+                                                  &ImagePathName, &RtlpPathDividers, &Position)))
+        {
+            ImagePathName.Length = Position + sizeof(WCHAR);
+        }
+
+        RequiredSize = ImagePathName.Length + DllNameOnly.Length +
+            (HasExtension ? 0 : RtlpDefaultExtension.Length) + sizeof(UNICODE_NULL);
+
+        if (RequiredSize >= UNICODE_STRING_MAX_BYTES)
+        {
+            Status = STATUS_NAME_TOO_LONG;
+        }
+        else
+        {
+            if (RequiredSize > CopyRealName.MaximumLength)
+            {
+                CopyRealName.Buffer = RtlpAllocateStringMemory(RequiredSize, TAG_USTR);
+                if (CopyRealName.Buffer == NULL)
+                    return STATUS_NO_MEMORY;
+                CopyRealName.MaximumLength = RequiredSize;
+            }
+            CopyRealName.Length = 0;
+            Status = RtlAppendUnicodeStringToString(&CopyRealName, &ImagePathName);
+            if NT_SUCCESS(Status)
```

Comment:
Uhm.

Reply:
Not sure what happened here...
This isnt python

Reply:
This is why it didnt bomb:
`#define NT_SUCCESS(Status)              (((NTSTATUS)(Status)) >= 0)`

Reply:
If you have `#define NT_SUCCESS(Status) (...)` then this type of construct works 😄